baseapp/baseapp.go

```diff
@@ -703,37 +704,35 @@ func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode) (*s
 		}
 
 		var (
-			msgEvents sdk.Events
-			msgResult *sdk.Result
-			msgFqName string
-			err       error
+			msgResult      *sdk.Result
+			msgEventAction string // name to use as value in event `message.action`
+			err            error
 		)
 
-		if svcMsg, ok := msg.(sdk.ServiceMsg); ok {
-			msgFqName = svcMsg.MethodName
-			handler := app.msgServiceRouter.Handler(msgFqName)
-			if handler == nil {
-				return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unrecognized message service method: %s; message index: %d", msgFqName, i)
-			}
-			msgResult, err = handler(ctx, svcMsg.Request)
-		} else {
+		if handler := app.msgServiceRouter.Handler(msg); handler != nil {
+			// ADR 031 request type routing
+			msgResult, err = handler(ctx, msg)
+			msgEventAction = sdk.MsgName(msg)
```

Comment:
type URL right?

Reply:
Let's keep proto MessageName to not introduce breaking changes.

Reply:
Yes, changed! See https://github.com/cosmos/cosmos-sdk/pull/9139#issuecomment-827503574 for reason.